### DISCUSSION TITLE
How use multiple cluster with same match route??

### DISCUSSION THREAD

**Discussion by cristyonathan** [dotnet/yarp#1481] [Q&A] [answered]
i want to  use yarp to forward my selected url to another selected destination url based on (header-key), but if any the same match url with different cluster the second match route for the last cluster that i regitered is doesnt redirect to selection destination.
 **can yarp use multiple cluster with same match route,? but will be differentiated based on header-key** 

lets say i have three json data like this
```json
[
  {
    "appName": "hello-world-app",
    "apis": [
      {
        "apiUrl": "/api/v1/test",
        "apiDestination": "/posts/3",
        "hostDestination": "https://jsonplaceholder.typicode.com/"
      }
    ]
  },
  {
    "appName": "my-app",
    "apis": [
      {
        "apiUrl": "/api/v1/test",
        "apiDestination": "/posts/1",
        "hostDestination": "https://jsonplaceholder.typicode.com/"
      },
      {
        "apiUrl": "/api/v2/test",
        "apiDestination": "/posts/2",
        "hostDestination": "https://jsonplaceholder.typicode.com/"
      }
    ]
  }
]
```

**theres is two data with same apiUrl (it will be the same match route but different cluster) but different appName (its will be unique header values for Apus-app header key)**
and this is my code 

```c#
            foreach (var app in apiByAppGroups)
            {
                var destinations = new Dictionary<string, DestinationConfig>(StringComparer.OrdinalIgnoreCase);
                foreach (var api in app.Apis)
                {
                    routes.Add(new RouteConfig
                    {
                        RouteId = api.ApiUrl,
                        ClusterId = app.AppName,
                        Match = new RouteMatch
                        {
                            Path = api.ApiUrl,
                            Headers = new List<RouteHeader>
                            {
                                new RouteHeader
                                {
                                    IsCaseSensitive = false,
                                    Name = "apus-app",
                                    Values = new List<string>
                                    {
                                        app.AppName
                                    }
                                }
                            }
                        },
                        Transforms = new List<Dictionary<string, string>>
                        {
                            new Dictionary<string, string>
                            {
                                { "PathSet", api.ApiDestination }
                            }
                    }
                    });
                    
                    destinations.Add(api.ApiDestination, new DestinationConfig()
                    {
                        Address = api.HostDestination
                    });
                }

                clusters.Add( new ClusterConfig
                {
                    ClusterId = app.AppName,
                    Destinations = destinations
                });
            }
```

**Comment by samsp-msft** [comment, marked as answer]
Each route will direct to one cluster. I think what you are trying to ask is can you have 2 routes that match the same URL, but differentiate based on a header in the request? That is possible, and then those two routes can point at different clusters.

**Reply by Tratcher** [reply]
The most specific match should win. If the paths are the same for two routes, then the headers will be used as a tie-breaker. Your setup seems correct for that. Can you give an example of a request and which route / cluster it ends up on?

Wait, I think I see your issue. You have `RouteId = api.ApiUrl,` and two routes with the same url, but RouteId needs to be unique. Consider making that `RouteId = app.AppName + api.ApiUrl,` or some other unique combination.

**Reply by cristyonathan** [reply]
yes thats the question i wanna ask, i have 2routes with same match URL, but differentiate based on a header in the request. but its still not working only the last route from app name  my-app could be accessed, the other one hello-world-apphello-world-app return 404

**Reply by Tratcher** [reply]
Can you show the incoming request headers? Here's a related issue with header routing:
https://github.com/microsoft/reverse-proxy/issues/1494

**Reply by cristyonathan** [reply]
thanks this issue already solved using change the RouteId with this RouteId = app.AppName + api.ApiUrl,
